File: packages/core/src/plugins/builtin/claude/history.ts
import { createReadStream } from 'node:fs';
import { access, constants, readdir } from 'node:fs/promises';
import { createInterface } from 'node:readline';
import path from 'node:path';
import type { ChatMessage, MessageContent, SkillFileEntry } from '@qlan-ro/mainframe-types';
import { getSessionJsonlPath } from './transcript.js';
import { resolveSkillPath } from './skill-path.js';
import { createChildLogger } from '../../../logger.js';
import {
  convertHistoryEntry,
  synthesizeUnknownCommandFromUserEntry,
  synthesizeSkillLoadedFromUserEntry,
} from './history-converters.js';
import {
  collectAgentProgressTools,
  collectSubagentToolResults,
  captureAgentIdMapping,
  collectSubagentAssistantBlocks,
  attachSubagentToolResults,
  injectAgentChildren,
} from './history-subagents.js';

const log = createChildLogger('claude:history');

export async function discoverSessionJsonlFiles(
  sessionId: string,
  projectPath: string,
): Promise<{ primaryPath: string; allFiles: string[]; subagentFiles: Set<string> }> {
  const { jsonlPath, projectDir } = getSessionJsonlPath(sessionId, projectPath);

  try {
    await access(jsonlPath, constants.R_OK);
  } catch {
    return { primaryPath: jsonlPath, allFiles: [], subagentFiles: new Set() };
  }

  const jsonlFiles = [jsonlPath];
  const subagentFiles = new Set<string>();

  // Scan sibling .jsonl files (sidechains) with matching sessionId
  try {
    const entries = await readdir(projectDir);
    for (const entry of entries) {
      if (!entry.endsWith('.jsonl') || entry === sessionId + '.jsonl') continue;
      const filePath = path.join(projectDir, entry);
      const stream = createReadStream(filePath);
      try {
        const rl = createInterface({ input: stream, crlfDelay: Infinity });
        for await (const line of rl) {
          if (!line.trim()) continue;
          const first = JSON.parse(line);
          if (first.sessionId === sessionId) jsonlFiles.push(filePath);
          break;
        }
      } catch {
        /* skip unreadable files */
      } finally {
        stream.destroy();
      }
    }
  } catch {
    /* directory read failed, proceed with primary only */
  }

  // Scan subagent JSONL files
  const subagentDir = path.join(projectDir, sessionId, 'subagents');
  try {
    const subEntries = await readdir(subagentDir);
    for (const entry of subEntries) {
      if (!entry.endsWith('.jsonl')) continue;
      const filePath = path.join(subagentDir, entry);
      jsonlFiles.push(filePath);
      subagentFiles.add(filePath);
    }
  } catch {
    /* no subagents directory or unreadable */
  }

  return { primaryPath: jsonlPath, allFiles: jsonlFiles, subagentFiles };
}

export async function loadHistory(sessionId: string, projectPath: string): Promise<ChatMessage[]> {
  const { allFiles: jsonlFiles, subagentFiles } = await discoverSessionJsonlFiles(sessionId, projectPath);
  if (jsonlFiles.length === 0) return [];

  const messages: ChatMessage[] = [];
  const agentTools = new Map<string, MessageContent[]>();
  // Map of toolUseId → tool_result block, populated from subagent JONLs
  const subagentToolResults = new Map<string, MessageContent & { type: 'tool_result' }>();
  const seenUuids = new Set<string>();
  // CLI 2.1.118+ subagent JSONLs omit parentToolUseID. The link is kept on the
  // parent's tool_result via toolUseResult.agentId. Build that map as we walk
  // the parent file so subagent-file processing can resolve the parent id.
  const agentIdToParentToolUseId = new Map<string, string>();

  for (const file of jsonlFiles) {
    const isSubagentFile = subagentFiles.has(file);
    const stream = createReadStream(file);
    try {
      const rl = createInterface({ input: stream, crlfDelay: Infinity });
      for await (const line of rl) {
        if (!line.trim()) continue;
        log.trace({ sessionId, file, line }, '[jsonl]');
        try {
          const entry = JSON.parse(line);

          // isMeta user messages carrying skill content are written to JSONL
          // only (never emitted over stream-json), so history replay is the
          // only chance to surface a SkillLoadedCard for these turns. Detect
          // and synthesize a system 'skill_loaded' message BEFORE the generic
          // isMeta filter drops them below.
          //
          // Skip subagent and sidechain entries: each subagent loads its own
          // skills and writes its own "Base directory for this skill:" entry.
          // Live mode hides those from the parent thread (subagent activity
          // surfaces only through agent_progress + the parent's Task tool_use),
          // so promoting them on replay creates ghost SkillLoadedCards that
          // never appeared during the live session.
          if (entry.isMeta === true && entry.type === 'user' && !isSubagentFile && entry.isSidechain !== true) {
            const synthesized = synthesizeSkillLoadedFromUserEntry(entry, sessionId);
            if (synthesized) {
              if (!seenUuids.has(synthesized.id)) {
                seenUuids.add(synthesized.id);
                messages.push(synthesized);
              }
              continue;
            }
          }

          if (entry.isMeta === true) continue;
          if (entry.isCompactSummary === true || entry.isVisibleInTranscriptOnly === true) continue;

          // Subagent JSONL files: extract tool_result and assistant blocks to
          // inline under the parent's Agent tool_use. These entries must NOT
          // appear as top-level chat messages — they're surfaced via the
          // injectAgentChildren / attachSubagentToolResults pipeline below.
          if (isSubagentFile) {
            collectSubagentToolResults(entry, subagentToolResults);
            collectSubagentAssistantBlocks(entry, agentTools, agentIdToParentToolUseId);
            continue;
          }

          // Capture agentId → parent tool_use_id mapping from the parent's
          // tool_result for any Task/Agent dispatch. discoverSessionJsonlFiles
          // returns the parent file before subagent files, so the map is built
          // by the time subagent processing needs it.
          if (entry.type === 'user') captureAgentIdMapping(entry, agentIdToParentToolUseId);

          // Sidechain entries are subagent activity (Task/Agent tool spawns its
          // own CLI session whose messages share our sessionId but live in a
          // sibling JSONL). The subagent's first user message is its dispatch
          // prompt — converting it would render a ghost user bubble in the
          // parent thread. Skill-loaded synthesis above runs first so user-typed
          // /skill invocations are preserved.
          if (entry.isSidechain === true) continue;

          // "Unknown command: /X" — CLI feedback for slash commands that don't
          // resolve. Split into invocation bubble + error pill on replay.
          if (entry.type === 'user') {
            const synthesized = synthesizeUnknownCommandFromUserEntry(entry, sessionId);
            if (synthesized) {
              for (const m of synthesized) {
                if (seenUuids.has(m.id)) continue;
                seenUuids.add(m.id);
                messages.push(m);
              }
              continue;
            }
          }

          // Collect tool_use blocks from agent_progress events
          if (entry.type === 'progress' && entry.data?.type === 'agent_progress') {
            collectAgentProgressTools(entry, agentTools);
            continue;
          }

          const msg = convertHistoryEntry(entry, sessionId);
          if (!msg) continue;

          // Deduplicate: primary JSONL is processed first, wins on conflicts
          if (seenUuids.has(msg.id)) continue;
          seenUuids.add(msg.id);
          messages.push(msg);
        } catch {
          // Skip malformed lines
        }
      }
    } finally {
      stream.destroy();
    }
  }

  // Inject collected subagent tool calls after their parent Agent tool_use blocks
  if (agentTools.size > 0) {
    injectAgentChildren(messages, agentTools);
  }

  // Attach tool_result data from subagent JONLs to the injected tool_use blocks
  if (subagentToolResults.size > 0) {
    attachSubagentToolResults(messages, subagentToolResults);
  }

  return messages;
}

export async function extractPlanFilePaths(sessionId: string, projectPath: string): Promise<string[]> {
  const { allFiles: jsonlFiles } = await discoverSessionJsonlFiles(sessionId, projectPath);
  if (jsonlFiles.length === 0) return [];

  const { projectDir } = getSessionJsonlPath(sessionId, projectPath);
  const planFiles: string[] = [];

  for (const file of jsonlFiles) {
    const stream = createReadStream(file);
    try {
      const rl = createInterface({ input: stream, crlfDelay: Infinity });
      for await (const line of rl) {
        if (!line.trim()) continue;
        try {
          const entry = JSON.parse(line);
          if (entry.type !== 'user') continue;
          const tur = entry.toolUseResult as Record<string, unknown> | undefined;
          if (typeof tur?.plan === 'string' && typeof tur?.filePath === 'string') {
            planFiles.push(path.resolve(projectDir, tur.filePath as string));
          }
        } catch {
          /* skip malformed */
        }
      }
    } finally {
      stream.destroy();
    }
  }

  return planFiles;
}

export async function extractSkillFilePaths(sessionId: string, projectPath: string): Promise<SkillFileEntry[]> {
  const { allFiles: jsonlFiles } = await discoverSessionJsonlFiles(sessionId, projectPath);
  if (jsonlFiles.length === 0) return [];

  const seen = new Set<string>();
  const cache = new Map<string, string>();
  const skillFiles: SkillFileEntry[] = [];
  const push = (name: string): void => {
    const trimmed = name.trim();
    if (!trimmed || seen.has(trimmed)) return;
    seen.add(trimmed);
    skillFiles.push({ path: resolveSkillPath(projectPath, trimmed, cache), displayName: trimmed });
  };

  for (const file of jsonlFiles) {
    const stream = createReadStream(file);
    try {
      const rl = createInterface({ input: stream, crlfDelay: Infinity });
      for await (const line of rl) {
        if (!line.trim()) continue;
        try {
          const entry = JSON.parse(line);
          if (entry.type !== 'assistant') continue;
          const content = entry.message?.content;
          if (!Array.isArray(content)) continue;
          for (const block of content) {
            if (block?.type === 'tool_use' && block.name === 'Skill') {
              const skill = (block.input as { skill?: unknown } | undefined)?.skill;
              if (typeof skill === 'string' && skill) push(skill);
            }
          }
        } catch {
          /* skip malformed */
        }
      }
    } finally {
      stream.destroy();
    }
  }

  return skillFiles;
}

File: packages/core-rs/crates/mainframe-adapter-claude/src/history.rs
//! Ported from `packages/core/src/plugins/builtin/claude/history.ts`.
//!
//! Reads Claude's own JSONL transcripts (under `~/.claude/projects/<encoded>/`)
//! to reconstruct a chat's message history for resume, including subagent
//! (task_group) inlining. Also extracts plan- and skill-file paths from a
//! session's transcripts.

use std::collections::{HashMap, HashSet};
use std::path::Path;

use dirs::home_dir;
use mainframe_types::chat::{ChatMessage, MessageContent};
use mainframe_types::context::SkillFileEntry;
use serde_json::Value;
use tokio::fs::File;
use tokio::io::{AsyncBufReadExt, BufReader, Lines};

use crate::history_converters::{
    convert_history_entry, synthesize_skill_loaded_from_user_entry,
    synthesize_unknown_command_from_user_entry,
};
use crate::history_subagents::{
    attach_subagent_tool_results, capture_agent_id_mapping, collect_agent_progress_tools,
    collect_subagent_assistant_blocks, collect_subagent_tool_results, inject_agent_children,
};
use crate::skill_path::resolve_skill_path;

fn is_strict_true(v: Option<&Value>) -> bool {
    matches!(v, Some(Value::Bool(true)))
}

/// CLI parity: replace every char NOT in `[a-zA-Z0-9-]` with '-' (keeps dashes).
fn encode_project_path(project_path: &str) -> String {
    project_path
        .chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || c == '-' {
                c
            } else {
                '-'
            }
        })
        .collect()
}

fn session_jsonl_path(session_id: &str, project_path: &str) -> (String, String) {
    let encoded = encode_project_path(project_path);
    let project_dir = home_dir()
        .unwrap_or_default()
        .join(".claude")
        .join("projects")
        .join(&encoded);
    let jsonl_path = project_dir.join(format!("{session_id}.jsonl"));
    (
        jsonl_path.to_string_lossy().to_string(),
        project_dir.to_string_lossy().to_string(),
    )
}

async fn open_lines(file_path: &str) -> Option<Lines<BufReader<File>>> {
    File::open(file_path)
        .await
        .ok()
        .map(|f| BufReader::new(f).lines())
}

pub struct DiscoveredFiles {
    pub primary_path: String,
    pub all_files: Vec<String>,
    pub subagent_files: HashSet<String>,
}

pub async fn discover_session_jsonl_files(session_id: &str, project_path: &str) -> DiscoveredFiles {
    let (jsonl_path, project_dir) = session_jsonl_path(session_id, project_path);

    if tokio::fs::metadata(&jsonl_path).await.is_err() {
        return DiscoveredFiles {
            primary_path: jsonl_path,
            all_files: Vec::new(),
            subagent_files: HashSet::new(),
        };
    }

    let mut jsonl_files = vec![jsonl_path.clone()];
    let mut subagent_files: HashSet<String> = HashSet::new();

    // Scan sibling .jsonl files (sidechains) with matching sessionId.
    let self_name = format!("{session_id}.jsonl");
    if let Ok(mut entries) = tokio::fs::read_dir(&project_dir).await {
        while let Ok(Some(entry)) = entries.next_entry().await {
            let name = entry.file_name().to_string_lossy().to_string();
            if !name.ends_with(".jsonl") || name == self_name {
                continue;
            }
            let file_path = Path::new(&project_dir)
                .join(&name)
                .to_string_lossy()
                .to_string();
            if let Some(mut lines) = open_lines(&file_path).await {
                while let Ok(Some(line)) = lines.next_line().await {
                    if line.trim().is_empty() {
                        continue;
                    }
                    if let Ok(first) = serde_json::from_str::<Value>(&line)
                        && first.get("sessionId").and_then(Value::as_str) == Some(session_id)
                    {
                        jsonl_files.push(file_path.clone());
                    }
                    break; // only the first non-empty line
                }
            }
        }
    }

    // Scan subagent JSONL files.
    let subagent_dir = Path::new(&project_dir).join(session_id).join("subagents");
    if let Ok(mut sub_entries) = tokio::fs::read_dir(&subagent_dir).await {
        while let Ok(Some(entry)) = sub_entries.next_entry().await {
            let name = entry.file_name().to_string_lossy().to_string();
            if !name.ends_with(".jsonl") {
                continue;
            }
            let file_path = subagent_dir.join(&name).to_string_lossy().to_string();
            jsonl_files.push(file_path.clone());
            subagent_files.insert(file_path);
        }
    }

    DiscoveredFiles {
        primary_path: jsonl_path,
        all_files: jsonl_files,
        subagent_files,
    }
}

pub async fn load_history(session_id: &str, project_path: &str) -> Vec<ChatMessage> {
    let discovered = discover_session_jsonl_files(session_id, project_path).await;
    if discovered.all_files.is_empty() {
        return Vec::new();
    }
    let subagent_files = &discovered.subagent_files;

    let mut messages: Vec<ChatMessage> = Vec::new();
    let mut agent_tools: HashMap<String, Vec<MessageContent>> = HashMap::new();
    let mut subagent_tool_results: HashMap<String, MessageContent> = HashMap::new();
    let mut seen_uuids: HashSet<String> = HashSet::new();
    // CLI 2.1.118+ subagent JSONLs omit parentToolUseID; the link lives on the
    // parent's tool_result via toolUseResult.agentId. Build that map while
    // walking the parent file so subagent processing can resolve the parent id.
    let mut agent_id_to_parent_tool_use_id: HashMap<String, String> = HashMap::new();

    for file in &discovered.all_files {
        let is_subagent_file = subagent_files.contains(file);
        let Some(mut lines) = open_lines(file).await else {
            // TODO(port): TS lets a mid-read stream error reject loadHistory; the
            // port skips an unreadable file (graceful) — files were just discovered.
            continue;
        };
        while let Ok(Some(line)) = lines.next_line().await {
            if line.trim().is_empty() {
                continue;
            }
            tracing::trace!(module = "claude:history", session_id = %session_id, file = %file, line = %line, "[jsonl]");

            let entry: Value = match serde_json::from_str(&line) {
                Ok(v) => v,
                Err(_) => continue, // skip malformed lines
            };

            // isMeta user messages carrying skill content are written to JSONL
            // only; synthesize a skill_loaded message before the isMeta filter.
            if is_strict_true(entry.get("isMeta"))
                && entry.get("type").and_then(Value::as_str) == Some("user")
                && !is_subagent_file
                && !is_strict_true(entry.get("isSidechain"))
                && let Some(synthesized) =
                    synthesize_skill_loaded_from_user_entry(&entry, session_id)
            {
                if !seen_uuids.contains(&synthesized.id) {
                    seen_uuids.insert(synthesized.id.clone());
                    messages.push(synthesized);
                }
                continue;
            }

            if is_strict_true(entry.get("isMeta")) {
                continue;
            }
            if is_strict_true(entry.get("isCompactSummary"))
                || is_strict_true(entry.get("isVisibleInTranscriptOnly"))
            {
                continue;
            }

            if is_subagent_file {
                collect_subagent_tool_results(&entry, &mut subagent_tool_results);
                collect_subagent_assistant_blocks(
                    &entry,
                    &mut agent_tools,
                    Some(&agent_id_to_parent_tool_use_id),
                );
                continue;
            }

            if entry.get("type").and_then(Value::as_str) == Some("user") {
                capture_agent_id_mapping(&entry, &mut agent_id_to_parent_tool_use_id);
            }

            if is_strict_true(entry.get("isSidechain")) {
                continue;
            }

            if entry.get("type").and_then(Value::as_str) == Some("user")
                && let Some(synthesized) =
                    synthesize_unknown_command_from_user_entry(&entry, session_id)
            {
                for m in synthesized {
                    if seen_uuids.contains(&m.id) {
                        continue;
                    }
                    seen_uuids.insert(m.id.clone());
                    messages.push(m);
                }
                continue;
            }

            if entry.get("type").and_then(Value::as_str) == Some("progress")
                && entry
                    .get("data")
                    .and_then(|d| d.get("type"))
                    .and_then(Value::as_str)
                    == Some("agent_progress")
            {
                collect_agent_progress_tools(&entry, &mut agent_tools);
                continue;
            }

            let msg = match convert_history_entry(&entry, session_id) {
                Some(m) => m,
                None => continue,
            };
            if seen_uuids.contains(&msg.id) {
                continue;
            }
            seen_uuids.insert(msg.id.clone());
            messages.push(msg);
        }
    }

    if !agent_tools.is_empty() {
        inject_agent_children(&mut messages, &agent_tools);
    }
    if !subagent_tool_results.is_empty() {
        attach_subagent_tool_results(&mut messages, &subagent_tool_results);
    }

    messages
}

pub async fn extract_plan_file_paths(session_id: &str, project_path: &str) -> Vec<String> {
    let discovered = discover_session_jsonl_files(session_id, project_path).await;
    if discovered.all_files.is_empty() {
        return Vec::new();
    }
    let (_, project_dir) = session_jsonl_path(session_id, project_path);
    let mut plan_files: Vec<String> = Vec::new();

    for file in &discovered.all_files {
        let Some(mut lines) = open_lines(file).await else {
            continue;
        };
        while let Ok(Some(line)) = lines.next_line().await {
            if line.trim().is_empty() {
                continue;
            }
            let entry: Value = match serde_json::from_str(&line) {
                Ok(v) => v,
                Err(_) => continue,
            };
            if entry.get("type").and_then(Value::as_str) != Some("user") {
                continue;
            }
            let tur = entry.get("toolUseResult");
            let plan_is_string = tur
                .and_then(|t| t.get("plan"))
                .map(Value::is_string)
                .unwrap_or(false);
            let file_path = tur.and_then(|t| t.get("filePath")).and_then(Value::as_str);
            if plan_is_string && let Some(fp) = file_path {
                plan_files.push(path_resolve(&project_dir, fp));
            }
        }
    }

    plan_files
}

pub async fn extract_skill_file_paths(session_id: &str, project_path: &str) -> Vec<SkillFileEntry> {
    let discovered = discover_session_jsonl_files(session_id, project_path).await;
    if discovered.all_files.is_empty() {
        return Vec::new();
    }

    let mut seen: HashSet<String> = HashSet::new();
    let mut cache: HashMap<String, String> = HashMap::new();
    let mut skill_files: Vec<SkillFileEntry> = Vec::new();

    for file in &discovered.all_files {
        let Some(mut lines) = open_lines(file).await else {
            continue;
        };
        while let Ok(Some(line)) = lines.next_line().await {
            if line.trim().is_empty() {
                continue;
            }
            let entry: Value = match serde_json::from_str(&line) {
                Ok(v) => v,
                Err(_) => continue,
            };
            if entry.get("type").and_then(Value::as_str) != Some("assistant") {
                continue;
            }
            let content = match entry
                .get("message")
                .and_then(|m| m.get("content"))
                .and_then(Value::as_array)
            {
                Some(c) => c,
                None => continue,
            };
            for block in content {
                if block.get("type").and_then(Value::as_str) == Some("tool_use")
                    && block.get("name").and_then(Value::as_str) == Some("Skill")
                    && let Some(skill) = block
                        .get("input")
                        .and_then(|i| i.get("skill"))
                        .and_then(Value::as_str)
                        .filter(|s| !s.is_empty())
                {
                    push_skill_file(skill, project_path, &mut seen, &mut cache, &mut skill_files);
                }
            }
        }
    }

    skill_files
}

fn push_skill_file(
    name: &str,
    project_path: &str,
    seen: &mut HashSet<String>,
    cache: &mut HashMap<String, String>,
    skill_files: &mut Vec<SkillFileEntry>,
) {
    let trimmed = name.trim();
    if trimmed.is_empty() || seen.contains(trimmed) {
        return;
    }
    seen.insert(trimmed.to_string());
    skill_files.push(SkillFileEntry {
        path: resolve_skill_path(Some(project_path), trimmed, Some(cache)),
        display_name: trimmed.to_string(),
    });
}

/// Lexical `path.resolve(base, p)` for unix paths (no filesystem access):
/// returns `p` when absolute, else `base/p`, collapsing `.`/`..` segments.
fn path_resolve(base: &str, p: &str) -> String {
    let combined = if Path::new(p).is_absolute() {
        p.to_string()
    } else {
        format!("{base}/{p}")
    };
    let mut stack: Vec<&str> = Vec::new();
    for seg in combined.split('/') {
        match seg {
            "" | "." => {}
            ".." => {
                stack.pop();
            }
            s => stack.push(s),
        }
    }
    format!("/{}", stack.join("/"))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn encode_keeps_dashes_replaces_other_metachars() {
        assert_eq!(
            encode_project_path("/Users/x/my_proj.v2"),
            "-Users-x-my-proj-v2"
        );
        // existing dashes are preserved
        assert_eq!(encode_project_path("a-b/c"), "a-b-c");
    }

    #[test]
    fn path_resolve_absolute_passthrough() {
        assert_eq!(path_resolve("/proj/dir", "/abs/plan.md"), "/abs/plan.md");
    }

    #[test]
    fn path_resolve_joins_relative() {
        assert_eq!(
            path_resolve("/proj/dir", "plans/x.md"),
            "/proj/dir/plans/x.md"
        );
    }

    #[test]
    fn path_resolve_collapses_dotdot() {
        assert_eq!(path_resolve("/proj/dir", "../plan.md"), "/proj/plan.md");
    }

    #[tokio::test]
    async fn missing_session_returns_empty() {
        assert!(
            load_history("no-such-session-xyz", "/tmp/no-such-project-xyz")
                .await
                .is_empty()
        );
        assert!(
            extract_plan_file_paths("no-such-session-xyz", "/tmp/no-such-project-xyz")
                .await
                .is_empty()
        );
        assert!(
            extract_skill_file_paths("no-such-session-xyz", "/tmp/no-such-project-xyz")
                .await
                .is_empty()
        );
    }
}

// PORT STATUS: src/plugins/builtin/claude/history.ts (285 lines)
// confidence: high
// todos: 1
// notes: createReadStream+readline → tokio BufReader::lines() (CRLF-stripped).
// getSessionJsonlPath's encode uses [^a-zA-Z0-9-] (KEEPS dashes) — distinct from
// external-session-paths::encode_path ([^a-zA-Z0-9]). path.resolve is a lexical
// unix-only resolver (collapses ./..). loadHistory threads seen_uuids/agent_tools/
// subagent maps exactly as the TS; strict `=== true` vs `!== true` checks mapped
// to is_strict_true. The 1 TODO(port): TS lets a mid-read stream error reject the
// promise; the port skips an unreadable file (graceful) instead — files are just-
// discovered so this is an unlikely race. No TS __tests__ file for history.ts
// (its behavior is covered via history-converters); sanity tests cover encode +
// path_resolve + the empty-session short-circuit.
